### PR TITLE
Curative remedial actions

### PR DESCRIPTION
**Pull request by baptsegu**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** 
#15

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Add curative remedial action into the RAO optimisation engine

**What is the current behavior?** *(You can also link to an open issue here)*
curative remedial actions are not taken into account in the RAO optimisation engine

**What is the new behavior (if this is a feature change)?**
take into account curative remedial action

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*
no

**Other information**
/!\ Does not handle the case where one generator or one PST is subject to several remedial actions (e.g. one preventive and one curative).

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/ClosedOptimisationRaoNames.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/ClosedOptimisationRaoNames.java]: 2019

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/ClosedOptimisationRaoUtil.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/ClosedOptimisationRaoUtil.java]: 2019

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/GeneratorRedispatchCostsFiller.java [hunk 3]
```diff
@@ -7,103 +7,104 @@
...
-    /**
-     * Check if the remedial action is a Redispatch remedial action (i.e. with only
-     * one remedial action element and redispatch)
-     */
-    private boolean isRedispatchRemedialAction(RemedialAction remedialAction) {
-        return remedialAction.getRemedialActionElements().size() == 1 &&
-                remedialAction.getRemedialActionElements().get(0) instanceof RedispatchRemedialActionElement;
-    }
+
+    private HashMap<Optional<Contingency>, List<RedispatchRemedialActionElement>> redispatchingRemedialActions;
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/GeneratorRedispatchCostsFiller.java]: Replace HashMap by Map. Concrete type is not needed in the memeber declaration

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/GeneratorRedispatchVariablesFiller.java [hunk 4]
```diff
@@ -7,58 +7,52 @@
...
 
-    /**
-     * Check if the remedial action is a Redispatch remedial action (i.e. with only
-     * one remedial action element and redispatch)
-     */
-    private boolean isRedispatchRemedialAction(RemedialAction remedialAction) {
-        return remedialAction.getRemedialActionElements().size() == 1 &&
-                remedialAction.getRemedialActionElements().get(0) instanceof RedispatchRemedialActionElement;
-    }
+    private HashMap<Optional<Contingency>, List<RedispatchRemedialActionElement>> redispatchingRemedialActions;
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/GeneratorRedispatchVariablesFiller.java]: Replace HashMap by Map. Concrete type is not needed in the memeber declaration

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/PstAngleImpactOnBranchFlowFiller.java [hunk 5]
```diff
@@ -7,67 +7,62 @@
...
 
-    /**
-     * Check if the remedial action is a PST remedial action (i.e. with only
-     * one remedial action element and PST)
-     */
-    private boolean isPstRemedialAction(RemedialAction remedialAction) {
-        return remedialAction.getRemedialActionElements().size() == 1 &&
-                remedialAction.getRemedialActionElements().get(0) instanceof PstElement;
-    }
+    private HashMap<Optional<Contingency>, List<PstElement>> pstRemedialActions;
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/PstAngleImpactOnBranchFlowFiller.java]: Replace HashMap by Map. Concrete type is not needed in the memeber declaration

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/PstAngleVariablesFiller.java [hunk 6]
```diff
@@ -7,66 +7,69 @@
...
 
-    /**
-     * Check if the remedial action is a PST remedial action (i.e. with only
-     * one remedial action element and PST)
-     */
-    private boolean isPstRemedialAction(RemedialAction remedialAction) {
-        return remedialAction.getRemedialActionElements().size() == 1 &&
-                remedialAction.getRemedialActionElements().get(0) instanceof PstElement;
-    }
+    private HashMap<Optional<Contingency>, List<PstElement>> pstRemedialActions;
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/PstAngleVariablesFiller.java]: Replace HashMap by Map. Concrete type is not needed in the memeber declaration

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/RedispatchEquilibriumConstraintFiller.java [hunk 7]
```diff
@@ -7,60 +7,61 @@
...
 
-    /**
-     * Check if the remedial action is a Redispatch remedial action (i.e. with only
-     * one remedial action element and redispatch)
-     */
-    private boolean isRedispatchRemedialAction(RemedialAction remedialAction) {
-        return remedialAction.getRemedialActionElements().size() == 1 &&
-                remedialAction.getRemedialActionElements().get(0) instanceof RedispatchRemedialActionElement;
-    }
+    private HashMap<Optional<Contingency>, List<RedispatchRemedialActionElement>> redispatchingRemedialActions;
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/RedispatchEquilibriumConstraintFiller.java]: Replace HashMap by Map. Concrete type is not needed in the memeber declaration

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/RedispatchImpactOnBranchFlowFiller.java [hunk 8]
```diff
@@ -7,67 +7,62 @@
...
-
-    /**
-     * Check if the remedial action is a Redispatch remedial action (i.e. with only
-     * one remedial action element and redispatch)
-     */
-    private boolean isRedispatchRemedialAction(RemedialAction remedialAction) {
-        return remedialAction.getRemedialActionElements().size() == 1 &&
-                remedialAction.getRemedialActionElements().get(0) instanceof RedispatchRemedialActionElement;
-    }
+    private HashMap<Optional<Contingency>, List<RedispatchRemedialActionElement>> redispatchingRemedialActions;
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/fillers/RedispatchImpactOnBranchFlowFiller.java]: Replace HashMap by Map. Concrete type is not needed in the member declaration

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/post_processors/PstElementResultsPostProcessor.java [hunk 9]
```diff
@@ -40,34 +52,50 @@
...
-                                            finalTapPosition
-                                    )
-                            )
-                    );
-                })
-                .collect(Collectors.toList());
-        result.getPreContingencyResult().getRemedialActionResults().addAll(remedialActionsResult);
+
+        //make map of pst remedial actions
+        HashMap<Optional<Contingency>, List<RemedialAction>> pstRemedialActions = new HashMap<>();
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/post_processors/PstElementResultsPostProcessor.java]: Replace HashMap by Map. Concrete type is not needed in variable declaration

File: ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/post_processors/RedispatchElementResultsPostProcessor.java [hunk 10]
```diff
@@ -18,63 +20,79 @@
...
-            })
-            .collect(Collectors.toList());
-        result.getPreContingencyResult().getRemedialActionResults().addAll(remedialActionsResult);
-    }
 
-    private boolean isRedispatchRemedialAction(RemedialAction remedialAction) {
-        return remedialAction.getRemedialActionElements().size() == 1 &&
-                remedialAction.getRemedialActionElements().get(0) instanceof RedispatchRemedialActionElement;
+        //make map of redispatching remedial actions
+        HashMap<Optional<Contingency>, List<RemedialAction>> redispatchRemedialActions = new HashMap<>();
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/main/java/com/farao_community/farao/closed_optimisation_rao/post_processors/RedispatchElementResultsPostProcessor.java]: Replace HashMap by Map. Concrete type is not needed in the variable declaration

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/FillersTestCase.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/FillersTestCase.java]: 2019

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/FillersTestCase.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+            assertTrue(areConstraintsPresent(filler.constraintsProvided()));
+        });
+    }
+
+    public void postProcessorsTest(List<String> postProcesors, RaoComputationResult raoComputationResult) {
+        ClosedOptimisationRaoParameters parameters = new ClosedOptimisationRaoParameters();
+        parameters.addAllPostProcessors(postProcesors);
+
+        // load processor <String = class name, OptimisationPostProcessor = proccessor object>
+        HashMap<String, OptimisationPostProcessor> proccessorsMap = new HashMap<>();
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/FillersTestCase.java]: Replace HashMap by Map. Concrete type is not needed in the variable declaration

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPConstraintMock.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPConstraintMock.java]: 2019

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPConstraintMock.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @author Baptiste Seguinot {@literal <baptiste.seguinot at rte-france.com>}
+ */
+public class MPConstraintMock extends MPConstraint {
+
+    private String name;
+    private double lb;
+    private double ub;
+    private HashMap<String, Double> coefficients;
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPConstraintMock.java]: Replace HashMap by Map. Concrete type is not needed in the member declaration

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPObjectiveMock.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPObjectiveMock.java]: 2019

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPObjectiveMock.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+import com.google.ortools.linearsolver.MPVariable;
+
+import java.util.HashMap;
+
+/**
+ * @author Marc Erkol {@literal <marc.erkol at rte-france.com>}
+ */
+public class MPObjectiveMock extends MPObjective {
+
+    private HashMap<String, Double> coefficients;
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPObjectiveMock.java]: Replace HashMap by Map. Concrete type is not needed in the member declaration

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPSolverMock.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPSolverMock.java]: 2019

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPVariableMock.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/MPVariableMock.java]: 2019

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java [hunk 19]
```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java]: 2019

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+
+    }
+
+    @Test
+    public void testcase2() {
+        CracFile cracFile = JsonCracFile.read(CracFile.class.getResourceAsStream("/5_3nodes_preContingency_PSTandRD_N-1.json"));
+        InputStream is = JsonClosedOptimisationRaoResultTest.class.getResourceAsStream("/5_3nodes_preContingency_PSTandRD_N-1.xiidm");
+        Network network = Importers.loadNetwork("/5_3nodes_preContingency_PSTandRD_N-1.xiidm", is);
+
+        HashMap<String, Object> data = new HashMap<>();
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java]: Replace HashMap by Map. Concrete type is not needed in the variable declaration

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+        HashMap<Pair<String, String>, Double> pstSensitivities = new HashMap<>();
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java]: same

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+        HashMap<Pair<String, String>, Double> generatorSensitivities = new HashMap<>();
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java]: same

File: ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+        HashMap<String, Double> referenceFlows = new HashMap<>();
```
Comment (murgeyseb) [inline ra-optimisation/closed-optimisation-rao/src/test/java/com/farao_community/farao/closed_optimisation_rao/OptimisationFillersTest.java]: same

**Comment by baptsegu** [comment]
Changes have been made.

MPSolver.infinity() is mocked in tests with Supermock.

**Pull request merged**